shell/browser/ui/file_dialog_gtk.cc

```diff
@@ -179,34 +220,60 @@ class FileChooserDialog {
   CHROMEG_CALLBACK_1(FileChooserDialog,
                      void,
                      OnFileDialogResponse,
-                     GtkWidget*,
+                     GtkFileChooser*,
                      int);
 
-  GtkWidget* dialog() const { return dialog_; }
+  GtkFileChooser* dialog() const { return dialog_; }
+  GModule* gtk_module() const { return gtk_module_; }
 
  private:
   void AddFilters(const Filters& filters);
 
   electron::NativeWindowViews* parent_;
   electron::UnresponsiveSuppressor unresponsive_suppressor_;
 
-  GtkWidget* dialog_;
+  GtkFileChooser* dialog_;
   GtkWidget* preview_;
+  GModule* gtk_module_;
 
   Filters filters_;
   std::unique_ptr<gin_helper::Promise<gin_helper::Dictionary>> save_promise_;
   std::unique_ptr<gin_helper::Promise<gin_helper::Dictionary>> open_promise_;
 
+  // Sets |dialog| as transient for |parent|, which will keep it on top and
+  // center it above |parent|. Do nothing if |parent| is nullptr.
+  void SetGtkTransientForAura(aura::Window* parent) {
+    if (!parent || !parent->GetHost())
+      return;
+
+    void (*dl_gtk_native_dialog_set_transient_for)(void*, GtkWindow*) = NULL;
+    g_module_symbol(
+        gtk_module_, "gtk_native_dialog_set_transient_for",
+        reinterpret_cast<void**>(&dl_gtk_native_dialog_set_transient_for));
+    dl_gtk_native_dialog_set_transient_for(
+        dialog_, reinterpret_cast<GtkWindow*>(parent));
```

Comment:
This cast looks bogus.  An `aura::Window` is not a `GtkWindow`, and I doubt it's in-memory layout is similar enough for this to even have a chance of working.

Reply:
Let me look into this to confirm either way

Reply:
One is a C++ class with a vtable, while the other is a C structure.

In the best case scenario, the `gtk_native_dialog_set_transient_for` call will fail on a `g_return_if_fail` check inside GTK from a runtime GType check (which should be evident from warnings on stderr).  Worst case, you cause a segfault.

Reply:
Weird I never hit that in my testing or compilation. I have an idea to get the underlying GtkWindow that I will confirm in #gtk tomorrow. But yea this is definitely not right at all, and I feel like I just totally missed something when I initially wrote that code a year ago.

```c
GApplication *app = g_application_get_default();
GtkWindow *parent = gtk_application_get_active_window(GTK_APPLICATION(app));
// do stuff with parent like set transient for.
```

Also this `SetGtkTransientForAura` is a hack imo because the regular `gtk::SetGtkTransientForAura` in `chrome/browser/ui/gtk/gtk_util.cc` has no concept of knowing about `GtkFileChooserNative` support. I dropped the X11/GdkWindow specific stuff from that function, and added back the necessary parts. I am a little confused as to why there is X11 code for setting `transient-for` on a `GtkFileChooserDialog` when the code could rely on GTK itself using `gtk_window_set_transient_for`. Does the electron project own that `gtk_util.cc` or does it belong to Chrome/Chromium? I really have no clue what files belong to whom if I am being honest.

Reply:
I suspect that there isn't an underlying GtkWindow for the aura::Window.  If there was, then old code would have called `gtk_window_set_transient_for` rather than twiddling X properties behind GTK's back.

Reply:
I had done some digging and found that an aura::Window (which I guess is a gfx:: NativeWindow on Linux) eventually was backed by a GtkWindow member variable called fake_window_, but maybe fake_window_ really is just fake.

Should I just remove the set transient call in the case of a GtkFileChooserNative?